src/k8s/request.js

```diff
@@ -407,26 +407,27 @@ const addNetworks = (vm, template, getSetting, networks) => {
 };
 
 const addCloudInit = (vm, defaultDisk, getSetting) => {
-  if (getSetting(CLOUD_INIT_KEY)) {
-    const existingCloudInitVolume = getCloudInitVolume(vm);
-    const cloudInit = new CloudInit({
-      volume: existingCloudInitVolume,
+  const existingCloudInitVolume = getCloudInitVolume(vm);
+  const cloudInit = new CloudInit({
+    volume: existingCloudInitVolume,
+  });
+
+  const vmHostname = getSetting(HOST_NAME_KEY);
+  if (getSetting(USE_CLOUD_INIT_CUSTOM_SCRIPT_KEY)) {
+    cloudInit.setUserData(getSetting(CLOUD_INIT_CUSTOM_SCRIPT_KEY));
+  } else {
+    cloudInit.setPredefinedUserData({
+      hostname: vmHostname,
+      sshAuthorizedKeys: getSetting(AUTHKEYS_KEY),
     });
+  }
 
-    if (getSetting(USE_CLOUD_INIT_CUSTOM_SCRIPT_KEY)) {
-      cloudInit.setUserData(getSetting(CLOUD_INIT_CUSTOM_SCRIPT_KEY));
-    } else {
-      cloudInit.setPredefinedUserData({
-        hostname: getSetting(HOST_NAME_KEY),
-        sshAuthorizedKeys: getSetting(AUTHKEYS_KEY),
-      });
-    }
+  vm.spec.template.spec.hostname = vmHostname || getSetting(NAME_KEY);
```

Comment:
Can we make it more generic? 

I would rather add a method to cloudInit called getHostname. Which would search for a hostname in the whole cloud-config, e.g.: `^hostname:\s+(\S+)`

this way we would also cover user defined scripts


Please see https://cloudinit.readthedocs.io/en/latest/topics/modules.html#set-hostname

Thinks to consider:
- script must start with  `#cloud-config` (we could add support for other types later)
- to take multipart config into account. and find the `#cloud-config` one (mm can there be mutliple ones of this type?).
- fqdn has a priority in cloud config - so we could parse the hostname out of it  (??)
- getHostname should probably call `build` to resolve the headers.

Reply:
I talked with @jelkosz about parsing user-provided cloud-init scripts while I was working on the feature and he advised that it's outside the scope of this particular bug. We can see about adding support for it in separate PR, though.

Reply:
ok can you please create an issue for that and/or continue on this?